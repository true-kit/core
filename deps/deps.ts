import { Deps } from './deps.types';

export function getDeps<
	D extends Deps<any>,
	P extends {deps?: D},
>(
	Target: (props: P) => JSX.Element,
	props: P,
): NonNullable<P['deps']> {
	return Object(props.deps);
}