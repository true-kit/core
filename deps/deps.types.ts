export type LikeComponent<P extends object> = (props: P) => JSX.Element;

export type DepsSpec = {
	[key:string]: object | undefined;
}

export type Deps<T extends DepsSpec> = {
	[K in keyof T]?: LikeComponent<any>;
}

export type DepsRegistry = {
	map: Map<LikeComponent<any>, Deps<any>> | null;
	overrides: null; // @todo
}