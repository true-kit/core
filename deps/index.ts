export * from './deps';
export * from './deps.types';