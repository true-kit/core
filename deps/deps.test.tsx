import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { Deps } from './deps.types';
import { getDeps } from './deps';

type IconProps = {
	name: string;
}

function BaseIcon(props: IconProps) {
	return <i className={props.name}/>
}

function EmIcon(props: IconProps) {
	return <em className={props.name}/>
}

type ButtonProps = {
	iconName: string;

	deps?: Deps<{
		Icon: IconProps;
	}>;
}

function IconButton(props: ButtonProps) {
	const deps = getDeps(IconButton, props);
	const {
		Icon = BaseIcon,
	} = deps;

	return (
		<button>
			<Icon name={props.iconName}/>
		</button>
	);
}

const root = document.createElement('div');

function render(fragment: JSX.Element) {
	ReactDOM.render(fragment, root,);
	return root.innerHTML;
}

it('deps: without context & inline', () => {
	expect(render(<IconButton iconName="ok" />)).toMatchSnapshot();
});

it('deps: inline', () => {
	expect(render(<IconButton iconName="inline" deps={{Icon: EmIcon}} />)).toMatchSnapshot();
});

it('deps: with inline', () => {
	expect(render(<IconButton iconName="inline" deps={{Icon: EmIcon}} />)).toMatchSnapshot();
});