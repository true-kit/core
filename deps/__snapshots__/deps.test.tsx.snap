// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deps: inline 1`] = `"<button><em class=\\"inline\\"></em></button>"`;

exports[`deps: with inline 1`] = `"<button><em class=\\"inline\\"></em></button>"`;

exports[`deps: without context & inline 1`] = `"<button><i class=\\"ok\\"></i></button>"`;
